### DISCUSSION TITLE
C++基础知识小讲堂(2)——RAII与三/五/零法则 – ChloroBlog

### DISCUSSION THREAD

**Discussion by Chlorie** [Chlorie/ChloroBlog#33] [Announcements]
https://chlorie.github.io/ChloroBlog/posts/2019-10-17/0-cpp-basics-2.html 

 能鸽善鹉者的领域

**Comment by c0nstexpr** [comment]
```好，很基础```

**Comment by tangm421** [comment]
> 在复制赋值运算符中我们对“自赋值”的情况做了特殊的处理，而对于移动赋值运算符，我们并不需要进行这种特殊处理。因为并没有办法把一个临时量赋值给自己，一个临时量是没法同时在等号两边出现的

还是得考虑加上对“自赋值”的判断吧？比如使用std::move将左值移动为右值引用再进行自赋值的情形：
```c++
std::string s1("123456");
s1 = std::move(s1);
```

**Comment by Chlorie** [comment]
> @tangm421
> > 在复制赋值运算符中我们对“自赋值”的情况做了特殊的处理，而对于移动赋值运算符，我们并不需要进行这种特殊处理。因为并没有办法把一个临时量赋值给自己，一个临时量是没法同时在等号两边出现的
> 
> 还是得考虑加上对“自赋值”的判断吧？比如使用std::move将左值移动为右值引用再进行自赋值的情形：
> ```c++
> std::string s1("123456");
> s1 = std::move(s1);
> ```

这个其实还是有点争议的（？）有的人觉得这种算是 use after move，有人觉得这种没问题。不过我最近写代码倒是都会判断移动自赋值的。

**Comment by tangm421** [comment]
> @Chlorie
> > @tangm421
> > > 在复制赋值运算符中我们对“自赋值”的情况做了特殊的处理，而对于移动赋值运算符，我们并不需要进行这种特殊处理。因为并没有办法把一个临时量赋值给自己，一个临时量是没法同时在等号两边出现的
> > 
> > 还是得考虑加上对“自赋值”的判断吧？比如使用std::move将左值移动为右值引用再进行自赋值的情形：
> > ```c++
> > std::string s1("123456");
> > s1 = std::move(s1);
> > ```
> 
> 这个其实还是有点争议的（？）有的人觉得这种算是 use after move，有人觉得这种没问题。不过我最近写代码倒是都会判断移动自赋值的。

看了下stl实现的std::string中，也是处理了，感觉最好还是都写上。毕竟只要是涉及到有堆内存分配的地方，不处理始终都有隐患。

![Snipaste_2021-04-21_16-34-26.png](https://ae02.alicdn.com/kf/H9178b87197204dcabee61fc0c3655643K.png)
